storage/watchable_store.go

```diff
@@ -241,6 +242,59 @@ func (s *watchableStore) syncWatchingsLoop() {
 	}
 }
 
+// RangeAllUnsynced ranges on all unsynced watchings
+// and returns all key-value pairs and next revision.
+func (s *watchableStore) RangeAllUnsynced() ([]storagepb.KeyValue, int64, error) {
+	totalLimit := 0
+	minRev, maxRev := int64(math.MaxInt64), int64(math.MinInt64)
+	for w := range s.unsynced {
+		if w.cur > 0 && w.cur <= s.store.compactMainRev {
+			log.Printf("storage: %v", ErrCompacted)
+			delete(s.unsynced, w)
+			continue
+		}
+		if w.cur > s.store.currentRev.main {
+			log.Printf("storage: %v", ErrFutureRev)
+			delete(s.unsynced, w)
+			continue
+		}
+		totalLimit += cap(w.ch) - len(w.ch)
+		if minRev >= w.cur {
+			minRev = w.cur
+		}
+		if maxRev <= w.cur {
+			maxRev = w.cur
+		}
+	}
+
+	min, max := newRevBytes(), newRevBytes()
+	revToBytes(revision{main: minRev}, min)
+	revToBytes(revision{main: maxRev, sub: maxRev}, max)
+
+	s.store.mu.Lock()
+	defer s.store.mu.Unlock()
+
+	tx := s.store.b.BatchTx()
+
+	tx.Lock()
+	defer tx.Unlock()
+
+	kvs := []storagepb.KeyValue{}
+
+	_, vs := tx.UnsafeRange(keyBucketName, min, max, int64(totalLimit))
```

Comment:
limit limits the number of key,values returned. I am not sure why do we need to limit it in this case.